### PR TITLE
Introducing stacks-devnet-js

### PR DESCRIPTION
**Pull request by lgalabru**
With this PR, we're turning the `clarinet integrate` feature into a node-api / n-api dynamic library that developers can use for spawning a devnet setup so full end to end integration can be written.

```typescript
const orchestrator = new StacksDevnetOrchestrator({
  path: "../../clariswap/Clarinet.toml",
  logs: true,
});
const network = new StacksTestnet({ url: orchestrator.getStacksNodeUrl() });
orchestrator.start();

var block = orchestrator.waitForStacksBlock();
console.log(`Genesis Block: ${JSON.stringify(block)}`);
  
const txOptions = {
      recipient: 'ST1SJ3DTE5DN7X54YDH5D64R3BCB6A2AG2ZQ8YPD5',
      amount: new BigNum(12345),
      senderKey: '753b7cc01a1a2e86221266a154af739463fce51219d97e4f856cd7200c3bd2a601',
      network,
      memo: 'test memo',
      nonce: new BigNum(0), // set a nonce manually if you don't want builder to fetch from a Stacks node
      fee: new BigNum(200), // set a tx fee if you don't want the builder to estimate
      anchorMode: AnchorMode.OnChainOnly
};
    
const transaction = await makeSTXTokenTransfer(txOptions);
    
// to see the raw serialized tx
const serializedTx = transaction.serialize().toString('hex');
    
// broadcasting transaction to the specified network
await broadcastTransaction(transaction, network);
    
block = orchestrator.waitForStacksBlock();
console.log(`Next Block: ${JSON.stringify(block)}`);

orchestrator.stop();

```

## Todos left
- [ ] documentation + examples

### CODE REVIEW

File: node-bindings/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+                            if let Some(c) = callback {
+                                c(&channel);
+                            }
+                            break;
+                        }
+                        DevnetCommand::Start(_) => break,
+                    }
+                }
+            } else {
+                // todo(ludo): Graceful termination.
```
Comment (obycode) [inline node-bindings/src/lib.rs:94]: Maybe we should have some logging here for now until the graceful termination is in place.

**Review by obycode — COMMENTED** [review]
I've gone through about half of the files and will finish the rest later today. Looking good! This will be really nice to have.

**Review by obycode — APPROVED** [review]
LGTM, just a few small nit-picks.

File: node-bindings/dist/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
+import { Block } from "./types";
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/dist/index.d.ts]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/8277c5112112d56f70d4c20a4e0f0dbb414c7c1a

File: node-bindings/lib/test.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { StacksDevnetOrchestrator } from "./index";
+
+// const devnet = new StacksDevnet({
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/lib/test.ts]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/4a74c51f592ac68fd6948d7818fa5ff18f9eb7d2

File: node-bindings/lib/test.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+//     config: {
+//         manifestPath: "/Users/ludovic/Coding/clarinet/clarinet-cli/examples/counter/Clarinet.toml",
+//     },
+//     logger: (msg) => console.log(msg)
+// });
+
+const devnet = new StacksDevnetOrchestrator({
+    path: "/Users/ludovic/Coding/clarinet/clarinet-cli/examples/counter/Clarinet.toml",
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/lib/test.ts]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/4a74c51f592ac68fd6948d7818fa5ff18f9eb7d2

File: node-bindings/package.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "stacks-devnet-js",
+  "version": "0.18.3",
+  "description": "stacks-devnet-js is a library for writing end to end tests for protocols interacting with the Stacks blockchain and the Bitcoin blockchain.",
+  "author": "Ludo Galabru",
+  "license": "ISC",
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/package.json]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/4a74c51f592ac68fd6948d7818fa5ff18f9eb7d2

File: node-bindings/src/lib.rs (hunk 1 above)
Comment (lgalabru) [reply node-bindings/src/lib.rs]: I removed that branch in https://github.com/hirosystems/clarinet/pull/157/commits/4a74c51f592ac68fd6948d7818fa5ff18f9eb7d2

File: node-bindings/src/serde/errors.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        NotImplemented(name: &'static str) {
+            description("Not Implemented")
+            display("Not Implemented: '{}'", name)
+        }
+        /// A JS exception was thrown
+        Js(throw: neon::result::Throw) {
+            description("JS exception")
+            display("JS exception")
+        }
+        // failed to convert something to f64
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/src/serde/errors.rs]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/669d94b2c183f1ae641d77a61ff4125eda5c4a66

File: node-bindings/src/serde/ser.rs [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+    cx: &'a mut C,
+    array: Handle<'j, JsArray>,
+}
+
+#[doc(hidden)]
+pub struct TupleVariantSerializer<'a, 'j, C: 'a>
+where
+    C: Context<'j>,
+{
+    outter_object: Handle<'j, JsObject>,
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply node-bindings/src/serde/ser.rs]: Addressed with https://github.com/hirosystems/clarinet/pull/157/commits/669d94b2c183f1ae641d77a61ff4125eda5c4a66

**Comment by lgalabru** [comment]
Thank you for your feedbacks @obycode! I'll go ahead, merge and generate a release `v0.19.0` so we can test the new CI, and test this whole integration end to end (including the `npm install stacks-devnet-js` process). 
We will probably follow up with a `v0.19.1` very soon, fixing the issues being observed.

**Pull request merged**